datafusion/physical-plan/src/sorts/partial_sort.rs

```diff
@@ -0,0 +1,992 @@
+    }
+}
+
+struct PartialSortStream {
+    /// The input plan
+    input: SendableRecordBatchStream,
+    /// Sort expressions
+    expr: Vec<PhysicalSortExpr>,
+    /// Length of
+    common_prefix_length: usize,
+    /// Used as a buffer for part of the input not ready for sort
+    input_batch: RecordBatch,
+    /// Fetch top N results
+    fetch: Option<usize>,
+    /// Whether the stream has finished returning all of its data or not
+    is_closed: bool,
+    /// Execution metrics
+    baseline_metrics: BaselineMetrics,
+}
+
+impl Stream for PartialSortStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let poll = self.poll_next_inner(cx);
+        self.baseline_metrics.record_poll(poll)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // we can't predict the size of incoming batches so re-use the size hint from the input
+        self.input.size_hint()
+    }
+}
+
+impl RecordBatchStream for PartialSortStream {
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+}
+
+impl PartialSortStream {
+    fn poll_next_inner(
+        self: &mut Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<RecordBatch>>> {
+        if self.is_closed {
+            assert_eq!(0, self.input_batch.num_rows());
+            return Poll::Ready(None);
+        }
+        let result = match ready!(self.input.poll_next_unpin(cx)) {
+            Some(Ok(batch)) => {
+                self.input_batch =
+                    concat_batches(&self.schema(), [&self.input_batch, &batch])?;
+                if let Some(slice_point) =
+                    self.get_slice_point(self.common_prefix_length, &self.input_batch)?
+                {
+                    self.emit(slice_point)
+                } else {
+                    Ok(RecordBatch::new_empty(self.schema()))
+                }
+            }
+            Some(Err(e)) => Err(e),
+            None => {
+                self.is_closed = true;
+                // once input is consumed, sort the rest of the inserted batches in input_batch
+                self.emit(self.input_batch.num_rows())
+            }
+        };
+
+        Poll::Ready(Some(result))
+    }
+
+    /// Sort input_batch upto the slice_point and emit the sorted batch
+    ///
+    /// If fetch is specified for PartialSortStream `emit` will limit
+    /// the last RecordBatch returned and will mark the stream as closed
+    fn emit(self: &mut Pin<&mut Self>, slice_point: usize) -> Result<RecordBatch> {
+        let original_num_rows = self.input_batch.num_rows();
+        let segment = self.input_batch.slice(0, slice_point);
+        self.input_batch = self
+            .input_batch
+            .slice(slice_point, self.input_batch.num_rows() - slice_point);
+
+        let result = sort_batch(&segment, &self.expr, self.fetch)?;
```

Comment:
You can comment on why the subtraction of` remaining_fetch - result.num_rows()` is safe. We can mention `sort_batch` internals here.

Reply:
I was thinking you could potentially make the sorting more efficient here by using only the last `len - prefix` keys (so there are fewer required comparisons), but that will only work when there is a single partition. Perhaps that could be a follow on optimization / special case as well